README.md
# VSCode Vector Aquarium

English / [日本語](https://qiita.com/le-nn/items/244c638b5ecebc54868b)

Make your vscode an aquarium and heal you !

<img src="image.gif" width="600px">

<img src="assets/secondary.png" width="600px">

## Install

Search ```vscode-vector-aquarium``` in your vscode extension tab or install from following url.

https://marketplace.visualstudio.com/items?itemName=le-nn.vscode-vector-aquarium

## Features

* When tapped, bait appears and fish will chase and eat it.
* When fish tapped, fish will diffuse and escape.
* Other creature will swim.
* Grouped fish swim in swarm

## How it works

Swarm of fish is implemented with boid algorithm.
https://en.wikipedia.org/wiki/Boids

Draw and movement logic is here.
https://github.com/le-nn/vscode-vector-aquarium/tree/main/src/webview/libs

## Usage

Open extension tab.

<img src="open.png" width="600px">

### View on secondary side bar

Enable secondary side bar from ```View > Appearance > Secondary Side Bar```.
Drag from primary side bar and drop Fish icon to secondary side bar in order to make side bar enabled.

![](./assets/appearance.png)
![](./assets/secondary-zoom.png)

## Setting

Press F1 and input following command.

```
vscode-vector-aquarium.config
```

And setting file will open.
Here is example of settings.
The ```fish``` field represents a swarm of fish.
One child of array is one swarm. Its child field represents one fish.
The setting example is as the field name.

```json
{
    "isFoodEnabled": true,
    "isRippleEnabled": true,
    "fish": [
        [
            {
                "color": "#3f51b5",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#3f51b5",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
        ],
        [
            {
                "color": "#2196f3",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#2196f3",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ]
    ],
    "jerryfish": [
        {
            "color": "#9c27b0",
            "scale": 1,
            "location": {
                "x": 340,
                "y": 120
            }
        },
        {
            "color": "#e91e63",
            "scale": 0.88,
            "location": {
                "x": 120,
                "y": 230
            }
        }
    ],
    "lophophorata": [
        {
            "color": "#3f51b5",
            "scale": 1,
            "location": {
                "x": 120,
                "y": 200
            }
        }
    ]
}

```

## License

Designed with ♥ by le-nn. Licensed under the MIT License.


CHANGELOG.md
# Change Log

All notable changes to the "vscode-vector-aquarium" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/defaultSettingsJson.ts
export const defaultSettings = `
{
    "isFoodEnabled": true,
    "isRippleEnabled": true,
    "fish": [
        [
            {
                "color": "#3f51b5",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#3f51b5",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#3f51b5",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#3f51b5",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#3f51b5",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#2196f3",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#2196f3",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#2196f3",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#2196f3",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#2196f3",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#00bcd4",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#00bcd4",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#00bcd4",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#00bcd4",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#00bcd4",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#009688",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#009688",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#009688",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#009688",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#009688",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#009688",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#4caf50",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#4caf50",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#4caf50",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#4caf50",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#4caf50",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#cddc39",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#cddc39",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#cddc39",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#cddc39",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#cddc39",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#cddc39",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#ffeb3b",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ffeb3b",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ffeb3b",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ffeb3b",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ffeb3b",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#ffc107",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ffc107",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ffc107",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ffc107",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ffc107",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#ff9800",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ff9800",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ff9800",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ff9800",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ff9800",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#ff5722",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ff5722",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ff5722",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ff5722",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ff5722",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#ff5722",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#f44336",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#f44336",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#f44336",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#f44336",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#f44336",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#e91e63",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#e91e63",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#e91e63",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#e91e63",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#e91e63",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ],
        [
            {
                "color": "#9c27b0",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#9c27b0",
                "scale": 0.8,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#9c27b0",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#9c27b0",
                "scale": 0.9,
                "location": {
                    "x": 0,
                    "y": 0
                }
            },
            {
                "color": "#9c27b0",
                "scale": 1,
                "location": {
                    "x": 0,
                    "y": 0
                }
            }
        ]
    ],
    "jerryfish": [
        {
            "color": "#9c27b0",
            "scale": 1,
            "location": {
                "x": 340,
                "y": 120
            }
        },
        {
            "color": "#e91e63",
            "scale": 0.88,
            "location": {
                "x": 120,
                "y": 230
            }
        },
        {
            "color": "#00bcd4",
            "scale": 0.79,
            "location": {
                "x": 160,
                "y": 330
            }
        },
        {
            "color": "#cddc39",
            "scale": 0.9,
            "location": {
                "x": 80,
                "y": 220
            }
        },
        {
            "color": "#ff5722",
            "scale": 1.2,
            "location": {
                "x": 96,
                "y": 150
            }
        }
    ],
    "lophophorata": [
        {
            "color": "#3f51b5",
            "scale": 1,
            "location": {
                "x": 120,
                "y": 200
            }
        }
    ]
}
`;